### ISSUE TITLE
Add support for Neo4j's mathematical functions.

### ISSUE THREAD

**Issue by aaramg** [neo4j-contrib/cypher-dsl#124]
Hi.
Please support Math functions in Cypher-DSL. Especially I need  **abs()** function. But cannot find it.
Also, I have a question. Do you have a feature to manually write and somehow concatenate some custom written cypher query to the existing builder? 
I mean, have I anyway to workaround, for example, this situation, if I need something to use but there is no implementation yet??

Also please write some documentation in Cypher-DSL doc about using procedures in the builder.

Also, I could not find normal documentation about these Neo4jTemplate methods if I am not able to use Cypher-DSL because there are missing functionalities:

```
	<T> List<T> findAll(String cypherQuery, Class<T> domainType);
	<T> List<T> findAll(String cypherQuery, Map<String, Object> parameters, Class<T> domainType);
```
I guess it is very important.
1. If we return multiple statements, do we need to create an appropriate class and put it in `domainType` to be able to parse the result in one object. If not, how? 
2. How these `Map<String, Object> parameters` work? Can I put in Object LocalDate, List or Map there for example? Or if I have class Filter with 'a' and 'b' fields,, can I put on Object value Filter type and then in the query use 'filter.a' or 'filter.b'. what symbol is used in front of the dynamic variable (i.e. '$' or smth else)?
3. And can we call "gds" or "apoc" procedures in this `cypherQuery ` ??? To be more precise, can we write in `cypherQuery ` any cypher query which is processable by the server?? Is this `cypherQuery ` work as if we wrote the same script in `@Query` annotation?

**Comment by michael-simons** [comment]
Hi @aaramg 

Thanks for your ticket.
Indeed, the math category is not yet as pre-defined functions. You can however use this 

```java
@Test // GH-124
void customFunctionsShouldWork() {

	Expression customFunctionCall = Cypher.call("abs").withArgs(Cypher.literalOf(-1)).asFunction();
	Statement statement = Cypher.returning(customFunctionCall).build();
	Assertions.assertThat(cypherRenderer.render(statement)).isEqualTo("RETURN abs(-1)");
}
```

idiom as a workaround in the mean time.

Why there is no implementation yet? We didn't need it for our internally generated queries in SDN 6 / SDN-RX yet. Just as simple as that.

> Also please write some documentation in Cypher-DSL doc about using procedures in the builder.

Can you please open a 2nd ticket for that?

> Also, I could not find normal documentation about these Neo4jTemplate methods if I am not able to use Cypher-DSL because there are missing functionalities:

What is a "normal" documentation? 

Anyway, I guess you speaking about

https://github.com/spring-projects/spring-data-neo4j/blob/master/src/main/java/org/springframework/data/neo4j/core/Neo4jOperations.java#L97

vs

https://github.com/spring-projects/spring-data-neo4j/blob/master/src/main/java/org/springframework/data/neo4j/core/Neo4jOperations.java#L129

They are equivalent.

The former will use the Cypher-DSL to render the statement and pass the rendered query along the engine.

However, the following 3 points are different questions

1. The domain type is a class annotated with Spring Data Neo4js `@Node` annotation, that is known to the object mapper. It has nothing todo with the Cypher-DSL. The manual is here https://docs.spring.io/spring-data/neo4j/docs/6.0.3/reference/html/
2. Each entry in `Map<String, Object> parameters` will be passed to the query execution engine. Each value in the map will be bound to a parameter with the same name. So an entry "foo" -> `LocalDate.of(2021,2,1)` will be bound to a Cypher-parameter with the name `$foo`. This is the name that must appear in the query. You either write it down or you use this `https://github.com/neo4j-contrib/cypher-dsl/blob/main/neo4j-cypher-dsl/src/main/java/org/neo4j/cypherdsl/core/Cypher.java#L199` to define an expression you can use within the Cypher-DSL.
3. I don't understand this. You can use all kinds of stored procedures (see https://github.com/neo4j-contrib/cypher-dsl/blob/main/neo4j-cypher-dsl/src/test/java/org/neo4j/cypherdsl/core/ProcedureCallsIT.java). The `@Query` annotation is again not part of this library here but of Spring Data Neo4j. It only takes a string. We have a ticket on SDN to evaluate ideas how to allow output from the Cypher-DSL there https://github.com/spring-projects/spring-data-neo4j/issues/1911 

So, we are gonna add the missing math functions, I think that makes sense for many people.

Please open a 2nd ticket about documenting custom calls.

**Comment by michael-simons** [comment]
I hope that helps with regard to custom functions http://neo4j-contrib.github.io/cypher-dsl/main/#callingarbitraryproceduresandfunctions

**Comment by aaramg** [comment]
@michael-simons thank you very much for your fast response and support.
The interesting part I guess a lot of people would be interested in is when there are multiple return parameters (f.ex some manually created aggregation).
Or smth existing:
For example:
```
CREATE (french:Cuisine {name:'French'})
CREATE (italian:Cuisine {name:'Italian'})
CREATE (indian:Cuisine {name:'Indian'})
CREATE (lebanese:Cuisine {name:'Lebanese'})
CREATE (portuguese:Cuisine {name:'Portuguese'})
CREATE (british:Cuisine {name:'British'})
CREATE (mauritian:Cuisine {name:'Mauritian'})

CREATE (zhen:Person {name: "Zhen"})
CREATE (praveena:Person {name: "Praveena"})
CREATE (michael:Person {name: "Michael"})
CREATE (arya:Person {name: "Arya"})
CREATE (karin:Person {name: "Karin"})

CREATE (praveena)-[:LIKES {score: 9}]->(indian)
CREATE (praveena)-[:LIKES {score: 7}]->(portuguese)
CREATE (praveena)-[:LIKES {score: 8}]->(british)
CREATE (praveena)-[:LIKES {score: 1}]->(mauritian)

CREATE (zhen)-[:LIKES {score: 10}]->(french)
CREATE (zhen)-[:LIKES {score: 6}]->(indian)
CREATE (zhen)-[:LIKES {score: 2}]->(british)

CREATE (michael)-[:LIKES {score: 8}]->(french)
CREATE (michael)-[:LIKES {score: 7}]->(italian)
CREATE (michael)-[:LIKES {score: 9}]->(indian)
CREATE (michael)-[:LIKES {score: 3}]->(portuguese)

CREATE (arya)-[:LIKES {score: 10}]->(lebanese)
CREATE (arya)-[:LIKES {score: 10}]->(italian)
CREATE (arya)-[:LIKES {score: 7}]->(portuguese)
CREATE (arya)-[:LIKES {score: 9}]->(mauritian)

CREATE (karin)-[:LIKES {score: 9}]->(lebanese)
CREATE (karin)-[:LIKES {score: 7}]->(italian)
CREATE (karin)-[:LIKES {score: 10}]->(portuguese)
```
How can I build this kind of query when I have return parameters: ` from, to, similarity` ?? (query below)
```
 MATCH (p1:Person {name: 'Michael'})-[likes1:LIKES]->(cuisine)
 MATCH (p2:Person)-[likes2:LIKES]->(cuisine) WHERE p2 <> p1
 RETURN p1.name AS from,
        p2.name AS to,
        gds.alpha.similarity.cosine(collect(likes1.score), collect(likes2.score)) AS similarity
 ORDER BY similarity DESC
```
Can I create a class with these fields and put in Neo4jTemplate#findAll as `Class<T>` ???
If no. How can I build this kind of query?? I brought this gds operation as an example but there can be a lot of aggregation calls that can return a lot of parameters, metrics, etc...
I have found @QueryResult in the old topics but in new docs cannot find it. https://docs.spring.io/spring-data/neo4j/docs/current/reference/html/#reference

Thanks in advance

**Comment by michael-simons** [comment]
Please note that we are underway adding the mentioned functions to the CypherDSL.

For the rest: This is Spring Data Neo4j. 
SDN 6 has no notion of `@QueryResult`. If the projections are not enough (please read that chapter) https://docs.spring.io/spring-data/neo4j/docs/current/reference/html/#projections, you should run a custom mapping. 

I am gonna add documentation for that topic to what we already have. You can track it here: https://github.com/spring-projects/spring-data-neo4j/issues/2135

Thanks for your understanding and support.

**Comment by michael-simons** [comment]
The set of all current available mathematical functions will be released with 2021.0 this month.
